### PR TITLE
net/portmapper: Add upnp

### PR DESCRIPTION
**Pull request by JulianKnodt**
Before adding the portmapper for UPnP, I found that it takes a significant amt of time (2s), to confirm that UPnP is present. The current probing method does not allow for any backoff, or easy reuse of clients. This attempts to refactor it into a separate `Prober`, which can track state. I'll also probably add some more to use it elsewhere, but wanted to put it up to see if the idea generally makes sense.

Output from prober:
```
=== RUN   TestClientProber
    probe_test.go:25: Got: {PCP:false PMP:false UPnP:true}, <nil>
--- PASS: TestClientProber (3.00s)
```

Output from `probe`
```
=== RUN   TestClientProbe
    portmapper_test.go:40: Got: {PCP:false PMP:false UPnP:false}, <nil>
    portmapper_test.go:40: Got: {PCP:false PMP:false UPnP:false}, <nil>
--- PASS: TestClientProbe (0.62s)
=== RUN   TestClientProbeThenMap
    portmapper_test.go:51: Probe: {PCP:false PMP:false UPnP:false}, <nil>
    portmapper_test.go:53: CreateOrGetMapping: invalid IPPort, no NAT mapping available: no port mapping services were found
--- PASS: TestClientProbeThenMap (0.26s)

```

Addresses #682 (Not really sure if it _fixes_ it though)

### CODE REVIEW

File: cmd/tailscaled/depaware.txt [hunk 1]
```diff
@@ -170,6 +176,9 @@ tailscale.com/cmd/tailscaled dependencies: (generated by github.com/tailscale/de
         golang.org/x/crypto/salsa20/salsa                            from golang.org/x/crypto/nacl/box+
         golang.org/x/net/bpf                                         from github.com/mdlayher/netlink+
         golang.org/x/net/dns/dnsmessage                              from net+
+        golang.org/x/net/html                                        from golang.org/x/net/html/charset
+        golang.org/x/net/html/atom                                   from golang.org/x/net/html
+        golang.org/x/net/html/charset                                from github.com/huin/goupnp
```
Comment (bradfitz) [inline cmd/tailscaled/depaware.txt:181]: We're very very memory constrained on iOS. I don't think we want to be pulling in large Unicode tables on iOS.

This dependency edge seems like a problem. Is it needed? Do we have to fork the goupnp package?

File: net/portmapper/portmapper.go [hunk 2]
```diff
@@ -433,127 +432,6 @@ func parsePMPResponse(pkt []byte) (res pmpResponse, ok bool) {
 	return res, true
 }
 
-type ProbeResult struct {
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:436]: let's not do this code movement. It's already in a portmapper file in a portmapper package. It's okay if it's a little long. 

If we do want to move things around, it should be in its own commit that doesn't add new code.

File: net/dns/resolver/neterr_other.go [hunk 3]
```diff
@@ -12,4 +12,4 @@ func networkIsUnreachable(err error) bool { return false }
 // packetWasTruncated returns true if err indicates truncation but the RecvFrom
 // that generated err was otherwise successful. It always returns false on this
 // platform.
-func packetWasTruncated(err error) bool   { return false }
+func packetWasTruncated(err error) bool { return false }
```
Comment (JulianKnodt) [inline net/dns/resolver/neterr_other.go:15]: When I run `go fmt ./...` from the root directory this change gets added but I don't need to include it.

File: net/portmapper/portmapper.go [hunk 4]
```diff
@@ -351,10 +363,58 @@ func (c *Client) CreateOrGetMapping(ctx context.Context) (external netaddr.IPPor
...
+	}
+	mpnp.external = netaddr.IPPortFrom(gw, newPort)
+	d := time.Duration(pmpMapLifetimeSec) * time.Second / 2
+	mpnp.useUntil = time.Now().Add(d)
+	mpnp.client = client
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.mapping = mpnp
+	c.localPort = newPort
+	return mpnp.external, nil
```
Comment (JulianKnodt) [inline net/portmapper/portmapper.go:417]: Not sure if this is the cleanest way to implement port mapping, but it works on my machine™. I don't know how to test it on other clients, and we would probably want to gate this so it doesn't automatically start getting called without more testing.

File: net/portmapper/portmapper.go (hunk 2 above)
Comment (JulianKnodt) [reply net/portmapper/portmapper.go]: I've moved the probe result back to the same place, but I deleted the whole probe function so I hope that's fine.

File: net/portmapper/portmapper_test.go [hunk 5]
```diff
@@ -32,12 +32,13 @@ func TestClientProbe(t *testing.T) {
...
-	for i := 0; i < 2; i++ {
+	c.NewProber(context.Background())
+	for i := 0; i < 30; i++ {
 		if i > 0 {
 			time.Sleep(100 * time.Millisecond)
 		}
-		res, err := c.Probe(context.Background())
-		t.Logf("Got: %+v, %v", res, err)
+		res, err := c.Prober.CurrentStatus()
+		t.Logf("Got(t=%dms): %+v, %v", i*100, res, err)
```
Comment (JulianKnodt) [inline net/portmapper/portmapper_test.go:41]: I notice that it discovers UPnP after 2.2 seconds on my machine, so I added the time in this test to make it clear when it finds it.

File: cmd/tailscaled/depaware.txt (hunk 1 above)
Comment (JulianKnodt) [reply cmd/tailscaled/depaware.txt]: I've forked the package, and removed the dependency on the charset. There is still more in the package, can work on trimming it down to something more reasonable.

File: net/upnp/LICENSE [hunk 6]
```diff
@@ -0,0 +1,23 @@
+Copyright (c) 2013, John Beisley <johnbeisleyuk@gmail.com>
```
Comment (DentonGentry) [inline net/upnp/LICENSE:1]: I imagine this should be forked in a new repository, like https://github.com/tailscale/upnp, rather than including it here.

Comment (JulianKnodt) [reply net/upnp/LICENSE:1]: I asked Brad about it, and I think it's probably best to keep it in the main one for now, we can fork it off later if necessary

File: net/portmapper/upnp.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+func (u *upnpMapping) isCurrent() bool                { return u.useUntil.After(time.Now()) }
+func (u *upnpMapping) validUntil() time.Time          { return u.useUntil }
+func (u *upnpMapping) externalIPPort() netaddr.IPPort { return u.external }
+func (u *upnpMapping) release() {
+	u.client.DeletePortMapping(context.Background(), "", u.external.Port(), "udp")
+}
+
+type upnpClient interface {
```
Comment (bradfitz) [inline net/portmapper/upnp.go:30]: Why is this an interface?

Go generally only uses interfaces for something you consume, not something you provide.

File: net/portmapper/upnp.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+		newExternalPort uint16,
+		newProtocol string,
+		newInternalPort uint16,
+		newInternalClient string,
+		newEnabled bool,
+		newPortMappingDescription string,
+		newLeaseDuration uint32,
+	) (err error)
+
+	DeletePortMapping(ctx context.Context, NewRemoteHost string, NewExternalPort uint16, NewProtocol string) error
```
Comment (bradfitz) [inline net/portmapper/upnp.go:46]: lowercase param names.

also: docs. but first I'd figure out how this should all be structured type/interface-wise.

File: net/portmapper/upnp.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+	GetStatusInfo(ctx context.Context) (status string, lastErr string, uptime uint32, err error)
```
Comment (bradfitz) [inline net/portmapper/upnp.go:47]: uint32 -> time.Duration for an uptime.

File: net/portmapper/upnp.go [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+	GetExternalIPAddress(ctx context.Context) (externalIPAddress string, err error)
+
+	RequestTermination(ctx context.Context) error
+	RequestConnection(ctx context.Context) error
+}
+
+func AddAnyPortMapping(
```
Comment (bradfitz) [inline net/portmapper/upnp.go:54]: too many parameters. Also, I thought you were going to make this a method some stateful type?

File: net/portmapper/probe.go [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+	"inet.af/netaddr"
+	"tailscale.com/net/netns"
+)
+
+type Prober struct {
+	// pause signals the probe to either pause temporarily (true), or stop entirely (false)
+	// to restart the probe, send another pause to it.
+	pause chan<- bool
+
+	PMP *ProbeSubResult
```
[the remark this answers is not in the document]
Comment (JulianKnodt) [reply net/portmapper/probe.go:20]: Added docs; they're non-nil, I removed the level of indirection and hopefully that makes it more clear

File: net/portmapper/upnp.go (hunk 7 above)
Comment (JulianKnodt) [reply net/portmapper/upnp.go:30]: This is to abstract over the upnp client provided by the fork of the goupnp client, which directly exposes each of the protocols. Should I move the interface into that package?

File: net/portmapper/upnp.go (hunk 8 above)
Comment (JulianKnodt) [reply net/portmapper/upnp.go:46]: These functions are just mapping those from the forked goupnp client package, which I think autogenerates them from documentation. I can lowercase the param names, but the interface directly maps them

File: net/portmapper/upnp.go (hunk 9 above)
Comment (JulianKnodt) [reply net/portmapper/upnp.go:47]: this is autogenerated, I can check to see if there's an easy to convert this into a time, but I don't think there would be

**Comment by JulianKnodt** [comment]
Closed in favor of tinier changes, see #2232

**Comment by JulianKnodt** [comment]
Closed in favor of tinier changes, see #2232
